go-chaos/internal/cluster.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 Camunda Services GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	v1 "k8s.io/api/apps/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"strings"
+	"time"
+)
+
+func (c K8Client) GetZeebeStatefulSet() (*v1.StatefulSet, error) {
+	namespace := c.GetCurrentNamespace()
+	ctx := context.TODO()
+
+	helmLabel := meta.LabelSelector{
+		MatchLabels: map[string]string{"app.kubernetes.io/name": "zeebe"},
+	}
+
+	statefulSets := c.Clientset.AppsV1().StatefulSets(namespace)
+	sfs, err := statefulSets.List(ctx, meta.ListOptions{LabelSelector: labels.Set(helmLabel.MatchLabels).String()})
+	if err != nil {
+		return nil, err
+	}
+	if len(sfs.Items) == 1 {
+		return &sfs.Items[0], nil
+	}
+	if len(sfs.Items) == 0 {
+		// On SaaS the StatefulSet is just named "zeebe" without any identifying labels
+		return statefulSets.Get(ctx, "zeebe", meta.GetOptions{})
+	}
+	return nil, errors.New("could not uniquely identify the stateful set for Zeebe")
+}
+
+// ScaleZeebeCluster Scales the StatefulSet for Zeebe. Waits until scaling is complete before returning the initial scale.
+func (c K8Client) ScaleZeebeCluster(replicas int) (int, error) {
+	namespace := c.GetCurrentNamespace()
+	ctx := context.TODO()
+	var initialReplicas int
+
+	sfs, err := c.GetZeebeStatefulSet()
+	if err != nil {
+		return 0, err
+	}
+	statefulSets := c.Clientset.AppsV1().StatefulSets(namespace)
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
:+1: could we use on more places I guess

Reply:
Yep :+1: It only added it here because some of the setup methods change resources in rapid succession where conflicts are much more likely to occur. But in general it's a good practice I think.